### PR TITLE
RIUI-312: Create Question

### PR DESCRIPTION
**Pull request by cksanders**
https://tools.hmcts.net/jira/browse/RIUI-312

Please be mindful that there is another PR that involves refactoring the API implementation.

### CODE REVIEW

File: api/cases/case.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 const sscsCaseTemplate = require('./sscsCase.template');
 const generateRequest = require('../lib/request');
+const generatePostRequest = require('../lib/postRequest');
```
Comment (mike-suggitt) [inline api/cases/case.js]: Could we maybe put this into the request helper? The request is only really setting headers and optionally adding a proxy so we could easily pass everything else through such as method and body etc

File: api/cases/case.js [hunk 2]
```diff
@@ -11,10 +12,32 @@ function getCaseEvents(caseId, userId, options, caseType = 'Benefit', jurisdicti
     return generateRequest(`${config.services.ccd_data_api}/caseworkers/${userId}/jurisdictions/${jurisdiction}/case-types/${caseType}/cases/${caseId}/events`, options)
 }
 
-function getCaseWithEvents(caseId, userId, options, caseType = 'Benefit', jurisdiction = 'SSCS') {
+function postHearing(caseId, userId, options, jurisdictionId = 'SSCS') {
```
Comment (mike-suggitt) [inline api/cases/case.js]: Could we put these functions into a seperate directory/module?

File: api/cases/case.js [hunk 3]
```diff
@@ -11,10 +12,32 @@ function getCaseEvents(caseId, userId, options, caseType = 'Benefit', jurisdicti
...
+        jurisdiction: jurisdictionId,
+        panel: [{identity_token: 'string', name: userId}],
+        start_date: (new Date()).toISOString()
+    };
+    
+    return generatePostRequest(`${config.services.coh_cor_api}/continuous-online-hearings`, options)
+        .then(hearing => hearing.online_hearing_id);
+ }
+ 
+function getHearingId(caseId, userId, options) {
```
Comment (mike-suggitt) [inline api/cases/case.js]: Is this right that if we can't find a COR for a case then we create one?

File: api/cases/case.js [hunk 4]
```diff
@@ -65,40 +88,54 @@ function caseFileReducer(caseId, caseFile) {
...
-    });
+    };
+    
+    getHearingId(caseId, userId, options)
+        .then(hearingId => getCaseWithEvents(caseId, userId, hearingId, options))
+        .then( ([caseData, events, questions])=> {
+            const schema = JSON.parse(JSON.stringify(sscsCaseTemplate));
+            
+            caseData.events = events != null ? events.map(e => reduceEvent(e)) : [];
+            caseData.draft_questions_to_appellant = questions && getDraftQuestions(questions.questions);
```
Comment (mike-suggitt) [inline api/cases/case.js]: Would we not be able to just use "questions" here? Would that cause problems later?

File: api/cases/case.js (hunk 3 above)
Comment (mike-suggitt) [reply api/cases/case.js]: why's that?

File: api/cases/case.js (hunk 2 above)
Comment (mike-suggitt) [reply api/cases/case.js]: Just seen these functions DO exist in a seperate routes module and guessing these are maybe a copy?

File: api/questions/create.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        .then(hearing => hearing.online_hearing_id)
+}
+
+function getHearingId(caseId, userId, headers) {
+    return generateRequest(`${config.services.coh_cor_api}/continuous-online-hearings?case_id=${caseId}`, {headers})
+        .then(hearing => hearing.online_hearings[0] ? hearing.online_hearings[0].online_hearing_id : postHearing(caseId, userId, headers));
+}
+
+function formatQuestion(body, userId) {
+    return {
```
Comment (mike-suggitt) [inline api/questions/create.js]: Are we defining a schema here? Seems like this should all come from the COR API?

File: api/cases/case.js (hunk 3 above)
Comment (alectronic0) [reply api/cases/case.js]: cause no one has the functionality to create a COH other than us atm
- not sscs
- not ccd

File: src/app/domain/case-viewer/components/questions-panel/questions-panel.component.html [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+                    </div>
+                    <div data-selector="draft-questions-list" *ngFor="let round of item.value; let i = index">
+                        <div *ngFor="let question of round; let j = index">
+                            <p data-selector="draft-questions-details" *ngIf="j === 0">You sent {{round.length}} questions to the appellant at {{question.state_datetime | date:'shortTime' | lowercase }} on {{question.state_datetime | date}}</p>
+                            <div data-selector="draft-questions-item" class="jui-questions__item">
+                                <h3 data-selector="questions-subject" class="jui-questions__subject"><a data-selector="questions-subject-link" routerLink="/viewcase/{{this.route.snapshot.params.case_id}}/questions/{{question.id}}">{{question.header}}</a>
+                                </h3>
+                                <p data-selector="questions-meta-data" class="jui-questions__meta">Sent by <span class="jui-questions__author">{{question.owner_reference}}</span> on {{question.state_datetime | date}} at {{ question.state_datetime | date:'shortTime' | lowercase}}</p>
```
Comment (mike-suggitt) [inline src/app/domain/case-viewer/components/questions-panel/questions-panel.component.html]: Will likely need to add `'+0000` timezone string for the dates to make sure they're UTC dates. This issue raised itself when running tests on jenkins and was fixed for the case list (table component)

File: src/app/routing/routing.module.ts [hunk 7]
```diff
@@ -22,8 +26,20 @@ const routes: Routes = [
         component: ViewCaseComponent
     },
     {
-        path: 'viewcase/:case_id/:section/:section_item_id',
+        path: 'viewcase/:case_id/casefile/:section_item_id',
```
Comment (mike-suggitt) [inline src/app/routing/routing.module.ts]: You could leave this route for now and simply place below the specified question routes for now

**Comment by mike-suggitt** [comment]
Completed review and looks ok with a few caveats:
 * Currently a failing unit test - `QuestionsPanelComponent when no create param is in the url When there is one round of draft questions to appellant should display two draft questions meta data`
* Additional PRs will clean up some of the implementation
* Question creation DOES NOT work with JS turned off. Personally i'd say this was a blocker (given previous conversations with UX) but happy to let @pawelkaczmarekuk choose

**Pull request merged**